## Importing libraries

In [1]:
import pandas as pd
import networkx as nx
import numpy as np
import matplotlib.pyplot as plt
import re

In [2]:
#import Dataframes scraped
listings = pd.read_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\listings_amsterdam.csv")
reviews = pd.read_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\reviews_list_amsterdam.csv")

In [3]:
listings.head()

listing_id                                              title  \
0        2818                 @ Daniel's (camera vista giardino)   
1     8778725                                                 []   
2    15458166                       Private room with canal view   
3    17632137  Modern 2P apt at historic Dappermarket | Zoo, ...   
4    20203218       Private room in Eastern Docklands, Amsterdam   

                                           toprow  \
0   Stanza privata in appartamento - Host: Daniel   
1                                              []   
2  Stanza privata in casa a schiera - Host: Malti   
3               Intero appartamento - Host: Guido   
4          Stanza privata in casa - Host: Carolyn   

                                            roominfo  \
0  ['2 ospiti', '1 camera da letto', '2 letti', '...   
1                                                 []   
2  ['2 ospiti', '1 camera da letto', '1 letto', '...   
3  ['2 ospiti', '1 camera da letto', '1 letto', '...   
4  ['2 ospiti', '1 camera da letto', '1 letto', '...   

                                          facilities rating  \
0  ['Wi-fi', 'Lavatrice', 'Cortile', 'Asciugacape...   4.88   
1                                                 []     []   
2  ['Accesso alla spiaggia (sul lungomare)', 'Cuc...   4.83   
3  ['Cucina', 'Wi-fi', 'TV', 'Lavatrice', 'Frigor...   4.36   
4  ['Lungomare', 'Wi-fi', 'Vasca da bagno', 'Gioc...   4.85   

                                    link      reviewnumber  
0      https://www.airbnb.it/rooms/2818?  (278 recensioni)  
1   https://www.airbnb.it/rooms/8778725?                []  
2  https://www.airbnb.it/rooms/15458166?   (78 recensioni)  
3  https://www.airbnb.it/rooms/17632137?   (25 recensioni)  
4  https://www.airbnb.it/rooms/20203218?   (26 recensioni)

In [4]:
reviews.head()

reviewer_id                                            reviews  \
0     39647118  Daniel è una persona molto socievole , ci siam...   
1     31905488  Ottima accoglienza, abbiamo concordato l'orari...   
2     28738000  L'appartamento è molto pulito. Daniel è molto ...   
3     32161750  Daniel è una persona estremamente gentile. L'a...   
4     33086171  Daniel is a perfect host: he is really kind, p...   

           date  listing_id  
0   aprile 2016        2818  
1  ottobre 2015        2818  
2   giugno 2015        2818  
3   luglio 2015        2818  
4  ottobre 2017        2818

## Data cleaning

In [5]:
listings[listings.reviewnumber == '[]']

listing_id title toprow roominfo facilities rating  \
1       8778725    []     []       []         []     []   
6        593101    []     []       []         []     []   
10     15014784    []     []       []         []     []   
23     26298886    []     []       []         []     []   
26       731834    []     []       []         []     []   
28      4242987    []     []       []         []     []   
39      9385505    []     []       []         []     []   
48     11215207    []     []       []         []     []   
60     11406872    []     []       []         []     []   
77     12749937    []     []       []         []     []   
94     27754860    []     []       []         []     []   
102    13722586    []     []       []         []     []   
130    19422892    []     []       []         []     []   
160    44110751    []     []       []         []     []   
165    19369338    []     []       []         []     []   

                                      link reviewnumber  
1     https://www.airbnb.it/rooms/8778725?           []  
6      https://www.airbnb.it/rooms/593101?           []  
10   https://www.airbnb.it/rooms/15014784?           []  
23   https://www.airbnb.it/rooms/26298886?           []  
26     https://www.airbnb.it/rooms/731834?           []  
28    https://www.airbnb.it/rooms/4242987?           []  
39    https://www.airbnb.it/rooms/9385505?           []  
48   https://www.airbnb.it/rooms/11215207?           []  
60   https://www.airbnb.it/rooms/11406872?           []  
77   https://www.airbnb.it/rooms/12749937?           []  
94   https://www.airbnb.it/rooms/27754860?           []  
102  https://www.airbnb.it/rooms/13722586?           []  
130  https://www.airbnb.it/rooms/19422892?           []  
160  https://www.airbnb.it/rooms/44110751?           []  
165  https://www.airbnb.it/rooms/19369338?           []

In [6]:
listings = listings[listings.reviewnumber != '[]']
listings.shape

(156, 8)

In [7]:
#Remove emoji
listings["title"]=listings["title"].str.replace('[^\w\s+&#@/:%.,_-]', '', flags=re.UNICODE)

In [8]:
# transform reviwnumber attribute in an integer attribute
for i in range(listings.shape[0]):
    
    num = int(re.findall(r"[0123456789]+",listings.iloc[i]["reviewnumber"])[0])
    listings.loc[i,"reviewnumber"] = num
    

In [9]:
listings.dropna(subset = ["listing_id"], inplace=True)
listings['listing_id'] = listings['listing_id'].apply(np.int64)

In [10]:
listings.head()

listing_id                                              title  \
0        2818                    @ Daniels camera vista giardino   
2    15458166                       Private room with canal view   
3    17632137  Modern 2P apt at historic Dappermarket  Zoo, park   
4    20203218       Private room in Eastern Docklands, Amsterdam   
5    26560199                              Sea side of Amsterdam   

                                           toprow  \
0   Stanza privata in appartamento - Host: Daniel   
2  Stanza privata in casa a schiera - Host: Malti   
3               Intero appartamento - Host: Guido   
4          Stanza privata in casa - Host: Carolyn   
5                Intero appartamento - Host: Tony   

                                            roominfo  \
0  ['2 ospiti', '1 camera da letto', '2 letti', '...   
2  ['2 ospiti', '1 camera da letto', '1 letto', '...   
3  ['2 ospiti', '1 camera da letto', '1 letto', '...   
4  ['2 ospiti', '1 camera da letto', '1 letto', '...   
5   ['3 ospiti', 'Monolocale', '2 letti', '1 bagno']   

                                          facilities rating  \
0  ['Wi-fi', 'Lavatrice', 'Cortile', 'Asciugacape...   4.88   
2  ['Accesso alla spiaggia (sul lungomare)', 'Cuc...   4.83   
3  ['Cucina', 'Wi-fi', 'TV', 'Lavatrice', 'Frigor...   4.36   
4  ['Lungomare', 'Wi-fi', 'Vasca da bagno', 'Gioc...   4.85   
5  ['Cucina', 'Wi-fi', 'Ascensore', 'Lavatrice', ...   4.73   

                                    link reviewnumber  
0      https://www.airbnb.it/rooms/2818?          278  
2  https://www.airbnb.it/rooms/15458166?           25  
3  https://www.airbnb.it/rooms/17632137?           26  
4  https://www.airbnb.it/rooms/20203218?           15  
5  https://www.airbnb.it/rooms/26560199?          682

In [11]:
listings.isnull().sum()

listing_id      0
title           0
toprow          0
roominfo        0
facilities      0
rating          0
link            0
reviewnumber    0
dtype: int64

In [12]:
listings = listings.dropna()

In [13]:
reviews.isnull().sum()

reviewer_id    0
reviews        8
date           0
listing_id     0
dtype: int64

In [14]:
reviews[reviews.reviews.isnull()]

reviewer_id reviews            date  listing_id
1233       786573     NaN   febbraio 2019    25310162
4810      2997094     NaN    ottobre 2018    15110159
4822     44199467     NaN     luglio 2018    15110159
6684     49005532     NaN     giugno 2018    16696921
7721    219747165     NaN     aprile 2019    16081933
9025    364408418     NaN  settembre 2020    16401807
9794    186980469     NaN     agosto 2018     2497346
9938     96948308     NaN     aprile 2017     2497346

In [15]:
reviews = reviews.dropna()

In [16]:
listings

listing_id                                              title  \
0          2818                    @ Daniels camera vista giardino   
2      15458166                       Private room with canal view   
3      17632137  Modern 2P apt at historic Dappermarket  Zoo, park   
4      20203218       Private room in Eastern Docklands, Amsterdam   
5      26560199                              Sea side of Amsterdam   
..          ...                                                ...   
166    21354405     Modern & beautiful apt. in OudWest with garden   
167    23747999                                      Rivierenbuurt   
168    26000561          Cosy apartment in the centre of Amsterdam   
169    33990484            Lovely bright apartment near Vondelpark   
170    34882977    Great apartment in Pijp area, with large garden   

                                             toprow  \
0     Stanza privata in appartamento - Host: Daniel   
2    Stanza privata in casa a schiera - Host: Malti   
3                 Intero appartamento - Host: Guido   
4            Stanza privata in casa - Host: Carolyn   
5                  Intero appartamento - Host: Tony   
..                                              ...   
166               Intero appartamento - Host: Lotte   
167            Intero appartamento - Host: Paskalle   
168                Intero appartamento - Host: Demi   
169               Intero appartamento - Host: Fleur   
170                Intero appartamento - Host: Asya   

                                              roominfo  \
0    ['2 ospiti', '1 camera da letto', '2 letti', '...   
2    ['2 ospiti', '1 camera da letto', '1 letto', '...   
3    ['2 ospiti', '1 camera da letto', '1 letto', '...   
4    ['2 ospiti', '1 camera da letto', '1 letto', '...   
5     ['3 ospiti', 'Monolocale', '2 letti', '1 bagno']   
..                                                 ...   
166  ['2 ospiti', '1 camera da letto', '1 letto', '...   
167  ['2 ospiti', '1 camera da letto', '1 letto', '...   
168  ['2 ospiti', '1 camera da letto', '1 letto', '...   
169  ['2 ospiti', '1 camera da letto', '1 letto', '...   
170  ['4 ospiti', '1 camera da letto', '2 letti', '...   

                                            facilities rating  \
0    ['Wi-fi', 'Lavatrice', 'Cortile', 'Asciugacape...   4.88   
2    ['Accesso alla spiaggia (sul lungomare)', 'Cuc...   4.83   
3    ['Cucina', 'Wi-fi', 'TV', 'Lavatrice', 'Frigor...   4.36   
4    ['Lungomare', 'Wi-fi', 'Vasca da bagno', 'Gioc...   4.85   
5    ['Cucina', 'Wi-fi', 'Ascensore', 'Lavatrice', ...   4.73   
..                                                 ...    ...   
166  ['Cucina', 'Wi-fi', 'TV', 'Lavatrice', 'Cortil...   4.90   
167  ['Cucina', 'Wi-fi', 'TV', 'Patio o balcone', '...   4.91   
168  ['Cucina', 'Wi-fi', 'TV', 'Lavatrice', 'Asciug...   4.88   
169  ['Cucina', 'Wi-fi', 'Animali domestici ammessi...   4.93   
170  ['Cucina', 'Wi-fi', 'TV', 'Lavatrice', 'Cortil...   4.55   

                                      link     reviewnumber  
0        https://www.airbnb.it/rooms/2818?              278  
2    https://www.airbnb.it/rooms/15458166?               25  
3    https://www.airbnb.it/rooms/17632137?               26  
4    https://www.airbnb.it/rooms/20203218?               15  
5    https://www.airbnb.it/rooms/26560199?              682  
..                                     ...              ...  
166  https://www.airbnb.it/rooms/21354405?  (10 recensioni)  
167  https://www.airbnb.it/rooms/23747999?  (11 recensioni)  
168  https://www.airbnb.it/rooms/26000561?  (16 recensioni)  
169  https://www.airbnb.it/rooms/33990484?  (29 recensioni)  
170  https://www.airbnb.it/rooms/34882977?  (11 recensioni)  

[156 rows x 8 columns]

In [17]:
reviews

reviewer_id                                            reviews  \
0         39647118  Daniel è una persona molto socievole , ci siam...   
1         31905488  Ottima accoglienza, abbiamo concordato l'orari...   
2         28738000  L'appartamento è molto pulito. Daniel è molto ...   
3         32161750  Daniel è una persona estremamente gentile. L'a...   
4         33086171  Daniel is a perfect host: he is really kind, p...   
...            ...                                                ...   
11389     79427219  Amazing location! A walk along the canal and y...   
11390     24123958  Asya’s apartment is spacious and in a great lo...   
11391      7443681  Fantastic location, large space well laid out....   
11392    238055005  A gem of an apartment, very stylish with a vie...   
11393      3043503  Asya's place was fantastic. Lovely style and d...   

               date  listing_id  
0       aprile 2016        2818  
1      ottobre 2015        2818  
2       giugno 2015        2818  
3       luglio 2015        2818  
4      ottobre 2017        2818  
...             ...         ...  
11389   agosto 2019    34882977  
11390   luglio 2019    34882977  
11391   luglio 2019    34882977  
11392   luglio 2019    34882977  
11393   luglio 2019    34882977  

[11386 rows x 4 columns]

# Create Network files

In [18]:
binetwork = pd.merge(reviews, listings, on='listing_id')
binetwork = binetwork[["reviewer_id","listing_id", "date"]]
binetwork

reviewer_id  listing_id          date
0         39647118        2818   aprile 2016
1         31905488        2818  ottobre 2015
2         28738000        2818   giugno 2015
3         32161750        2818   luglio 2015
4         33086171        2818  ottobre 2017
...            ...         ...           ...
11381     79427219    34882977   agosto 2019
11382     24123958    34882977   luglio 2019
11383      7443681    34882977   luglio 2019
11384    238055005    34882977   luglio 2019
11385      3043503    34882977   luglio 2019

[11386 rows x 3 columns]

In [19]:
# We choice as weight of edges the number of reviews that a reviewer makes of a listings
weighted_net = binetwork.groupby(['reviewer_id','listing_id'])[['date']].count().reset_index()
weighted_net = weighted_net.rename(columns={'date': 'weight'})

In [20]:
weighted_net

reviewer_id  listing_id  weight
0             5567    12475352       1
1            10952        2818       1
2            11358    14622016       1
3            11869        2818       1
4            12798        2818       1
...            ...         ...     ...
11260    398863768    29901220       1
11261    401396332    18950531       1
11262    402408329    29901220       2
11263    404501715    29776071       1
11264    404864960    23575414       1

[11265 rows x 3 columns]

In [21]:
len(binetwork.reviewer_id.unique())

11110

We have 11265 edges but only 11110 reviewer, so there many reviewer who commented more listings

In [24]:
binetwork.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\temporal_binetwork_amsterdam.csv", index = False)

In [25]:
weighted_net.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\weighted_net_amsterdam.csv", index = False)

###### Create a snapshots network

In [22]:
binetwork.date.unique()

array(['aprile 2016', 'ottobre 2015', 'giugno 2015', 'luglio 2015',
       'ottobre 2017', 'gennaio 2017', 'luglio 2014', 'giugno 2012',
       'novembre 2009', 'febbraio 2020', 'novembre 2019', 'ottobre 2019',
       'settembre 2019', 'agosto 2019', 'luglio 2019', 'giugno 2019',
       'maggio 2019', 'aprile 2019', 'marzo 2019', 'gennaio 2019',
       'dicembre 2018', 'novembre 2018', 'ottobre 2018', 'settembre 2018',
       'agosto 2018', 'luglio 2018', 'giugno 2018', 'maggio 2018',
       'aprile 2018', 'marzo 2018', 'gennaio 2018', 'novembre 2017',
       'settembre 2017', 'agosto 2017', 'luglio 2017', 'giugno 2017',
       'maggio 2017', 'aprile 2017', 'marzo 2017', 'febbraio 2017',
       'dicembre 2016', 'novembre 2016', 'ottobre 2016', 'settembre 2016',
       'agosto 2016', 'luglio 2016', 'giugno 2016', 'maggio 2016',
       'marzo 2016', 'gennaio 2016', 'dicembre 2015', 'novembre 2015',
       'settembre 2015', 'agosto 2015', 'maggio 2015', 'aprile 2015',
       'marzo 2015',

In [23]:
for i in range(binetwork.shape[0]):
    num = (re.findall(r"[0123456789]+",binetwork.iloc[i]["date"])[0])
    binetwork.loc[i,"date"] = num

In [24]:
binetwork

reviewer_id  listing_id  date
0         39647118        2818  2016
1         31905488        2818  2015
2         28738000        2818  2015
3         32161750        2818  2015
4         33086171        2818  2017
...            ...         ...   ...
11381     79427219    34882977  2019
11382     24123958    34882977  2019
11383      7443681    34882977  2019
11384    238055005    34882977  2019
11385      3043503    34882977  2019

[11386 rows x 3 columns]

In [25]:
list(binetwork.date.unique())

['2016',
 '2015',
 '2017',
 '2014',
 '2012',
 '2009',
 '2020',
 '2019',
 '2018',
 '2013',
 '2011',
 '2010',
 '2021']

In [26]:
grouped = binetwork.groupby(binetwork.date)
snapshot2009 = grouped.get_group('2009') 
snapshot2010 = grouped.get_group('2010') 
snapshot2011 = grouped.get_group('2011') 
snapshot2012 = grouped.get_group('2012') 
snapshot2013 = grouped.get_group('2013')
snapshot2014 = grouped.get_group('2014') 
snapshot2015 = grouped.get_group('2015') 
snapshot2016 = grouped.get_group('2016') 
snapshot2017 = grouped.get_group('2017')
snapshot2018 = grouped.get_group('2018')
snapshot2019 = grouped.get_group('2019') 
snapshot2020 = grouped.get_group('2020') 
snapshot2021 = grouped.get_group('2021') 

In [27]:
snapshot2020.head()

reviewer_id  listing_id  date
9      330320508        2818  2020
381      2931776    20203218  2020
407     55946151    26560199  2020
408     28608550    26560199  2020
440     39503718      654931  2020

In [29]:
def weightsnap(snap_df):
    snap_df = snap_df.groupby(['reviewer_id','listing_id'])[['date']].count().reset_index()
    return snap_df.rename(columns={'date': 'weight'}) 

In [30]:
snapshot2009 = weightsnap(snapshot2009)
snapshot2010 = weightsnap(snapshot2010)
snapshot2011 = weightsnap(snapshot2011)
snapshot2012 = weightsnap(snapshot2012)
snapshot2013 = weightsnap(snapshot2013)
snapshot2014 = weightsnap(snapshot2014)
snapshot2015 = weightsnap(snapshot2015)
snapshot2016 = weightsnap(snapshot2016)
snapshot2017 = weightsnap(snapshot2017)
snapshot2018 = weightsnap(snapshot2018)
snapshot2019 = weightsnap(snapshot2019)
snapshot2020 = weightsnap(snapshot2020)
snapshot2021 = weightsnap(snapshot2021)

In [31]:
snapshot2013.head()

reviewer_id  listing_id  weight
0        46875      730916       1
1       115275      759060       1
2       136311      685990       1
3       139554      759060       1
4       359832      730916       1

In [32]:
snapshot2009.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2009_amsterdam.csv", index = False)
snapshot2010.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2010_amsterdam.csv", index = False)
snapshot2011.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2011_amsterdam.csv", index = False)
snapshot2012.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2012_amsterdam.csv", index = False)
snapshot2013.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2013_amsterdam.csv", index = False)
snapshot2014.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2014_amsterdam.csv", index = False)
snapshot2015.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2015_amsterdam.csv", index = False)
snapshot2016.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2016_amsterdam.csv", index = False)
snapshot2017.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2017_amsterdam.csv", index = False)
snapshot2018.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2018_amsterdam.csv", index = False)
snapshot2019.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2019_amsterdam.csv", index = False)
snapshot2020.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2020_amsterdam.csv", index = False)
snapshot2021.to_csv(r"C:\Users\Gianmarco\Dropbox\Il mio PC (DESKTOP-DM1F9TN)\Desktop\SocialNetworkAnalysis\progetto\Data\snapshot2021_amsterdam.csv", index = False)